primitives/runtime/src/traits.rs

```diff
@@ -1252,38 +1252,71 @@ impl<'a> codec::Input for TrailingZeroInput<'a> {
 	fn try_from_account(a: &AccountId) -> Option<Self> {
 		Self::try_from_sub_account::<()>(a).map(|x| x.0)
 	}
 
-	/// Convert this value amalgamated with the a secondary "sub" value into an account ID. This is
-	/// infallible.
+	/// Convert this value amalgamated with the a secondary "sub" value into an account ID,
+	/// truncating any unused bytes. This is infallible.
 	///
 	/// NOTE: The account IDs from this and from `into_account` are *not* guaranteed to be distinct
 	/// for any given value of `self`, nor are different invocations to this with different types
 	/// `T`. For example, the following will all encode to the same account ID value:
 	/// - `self.into_sub_account(0u32)`
 	/// - `self.into_sub_account(vec![0u8; 0])`
 	/// - `self.into_account()`
-	fn into_sub_account<S: Encode>(&self, sub: S) -> AccountId;
+	///
+	/// Also, if the seed provided to this function is greater than the number of bytes which fit
+	/// into this `AccountId` type, then it will lead to truncation of the seed, and potentially
+	/// non-unique accounts.
+	fn into_sub_account_truncating<S: Encode>(&self, sub: S) -> AccountId;
+
+	/// Same as `into_sub_account`, but ensuring that all bytes of the account's seed are used when
+	/// generating an account. This can help guarantee that different accounts are unique, besides
+	/// types which encode the same as noted above.
+	fn try_into_sub_account<S: Encode>(&self, sub: S) -> Option<AccountId>;
 
 	/// Try to convert an account ID into this type. Might not succeed.
 	fn try_from_sub_account<S: Decode>(x: &AccountId) -> Option<(Self, S)>;
 }
 
-/// Format is TYPE_ID ++ encode(parachain ID) ++ 00.... where 00... is indefinite trailing zeroes to
+/// Format is TYPE_ID ++ encode(sub-seed) ++ 00.... where 00... is indefinite trailing zeroes to
 /// fill AccountId.
-impl<T: Encode + Decode, Id: Encode + Decode + TypeId> AccountIdConversion<T> for Id {
-	fn into_sub_account<S: Encode>(&self, sub: S) -> T {
+impl<T: Encode + Decode + MaxEncodedLen, Id: Encode + Decode + TypeId> AccountIdConversion<T>
+	for Id
+{
+	// Take the `sub` seed, and put as much of it as possible into the generated account, but
+	// allowing truncation of the seed if it would not fit into the account id in full. This can
+	// lead to two different `sub` seeds with the same account generated.
+	fn into_sub_account_truncating<S: Encode>(&self, sub: S) -> T {
 		(Id::TYPE_ID, self, sub)
 			.using_encoded(|b| T::decode(&mut TrailingZeroInput(b)))
-			.expect("`AccountId` type is never greater than 32 bytes; qed")
+			.expect("All byte sequences are valid `AccountIds`; qed")
+	}
+
+	// Same as `into_sub_account_truncating`, but returns `None` if any bytes would be truncated.
+	fn try_into_sub_account<S: Encode>(&self, sub: S) -> Option<T> {
+		(Id::TYPE_ID, self, sub).using_encoded(|b| {
+			if b.len() > T::max_encoded_len() {
+				return None
+			};
+			let account = T::decode(&mut TrailingZeroInput(b))
```

Comment:
IMO the check above is wrong, because if you for example have an enum `max_encoded_len` returns you the length of the biggest enum variant. This is wrong here, just imagine you are decoding the shortest variant and then you would not have consumed all bytes. 

If you want to make it correct, you need to extend `TrailingZeroInput` to be able to check if the given input was consumed completely after having the type decoded.

Reply:
The `AccountId` type has constant length, but in general I see that it does not work for all `T`.  
So my proposal of adding a `const_encoded_len` could also fix this, right?

Reply:
How do you know that the `AccountId` has a constant length? `AccountId` is some generic value that could be anything (one of the reasons for this pr). Just because Polkadot uses `AccountId32`, it doesn't mean that everyone uses this or is required to use this.

Reply:
Yes therefore we could ad a `FixEncodedLen` trait analogous to `MaxEncodedLen` that requires a type to always have the same encoding len.

Reply:
+1

Reply:
I did something like what @kianenigma suggested. I check the length of the bytes used for the seed, and check that the length of bytes of the account is greater or equal to, otherwise, we know some bytes were truncated.

Reply:
Thanks. I also tried my luck on this but could not create a test that the function actually returns `Some()`. The `AccountId` type is u64, so an overflow is very likely anyway.

Reply:
You can just create a mock environment where account id is 32 bytes?